File: src/python/day06.py
# Input:
example_input = """
123 328  51 64 
 45 64  387 23 
  6 98  215 314
*   +   *   +  
"""


def parse_input(input: str) -> tuple[list[list[int]], list[str]]:
    res = input.strip().split("\n")
    numbers: list[list[int]] = []
    for line in res[:-1]:
        split_line = [s for s in line.split(" ") if s != ""]
        for i, num_str in enumerate(split_line):
            if len(numbers) <= i:
                numbers.append([])
            numbers[i].append(int(num_str))
    operation_line = res[-1]
    operations = [s for s in operation_line.split(" ") if s != ""]
    return numbers, operations


def calculate(numbers, operations) -> int:
    total = 0
    for nums, op in zip(numbers, operations):
        if op == "*":
            res = 1
            for n in nums:
                res *= n
            total += res
        elif op == "+":
            total += sum(nums)
    return total


def parse_input_column(input: str) -> tuple[list[list[int]], list[str]]:
    res = input.strip().split("\n")
    numbers: list[list[int]] = []
    for line in res[:-1]:
        split_line = list(line)
        for i, char in enumerate(split_line):
            if char != " ":
                while len(numbers) <= i:
                    numbers.append([])
                numbers[i].append(int(char))
    joined_numbers: list[list[int]] = [[]]
    for digit_list in numbers:
        if digit_list == []:
            joined_numbers.append([])
            continue
        number = 0
        for digit in digit_list:
            number = number * 10 + digit
        joined_numbers[-1].append(number)
    operation_line = res[-1]
    operations = [s for s in operation_line.split(" ") if s != ""]
    operations = list(operations)
    return joined_numbers, operations


input = open("input/day06.txt", "r").read()
# input = example_input.strip()
numbers, operations = parse_input(input)
print(calculate(numbers, operations))
numbers, operations = parse_input_column(input)
print(calculate(numbers, operations))

File: src/rust/day06.rs
enum Operation {
    Multiply,
    Add,
}

impl Operation {
    fn from_str(op: &str) -> Self {
        match op {
            "*" => Self::Multiply,
            "+" => Self::Add,
            _ => panic!("Invalid operation"),
        }
    }
}

fn parse_input_columns(input: &str) -> (Vec<Vec<u64>>, Vec<Operation>) {
    let cols = input
        .lines()
        .take_while(|l| !l.starts_with(['*', '+']))
        .flat_map(|line| line.split_whitespace().enumerate())
        .fold(Vec::new(), |mut acc, (i, num_str)| {
            if i >= acc.len() {
                acc.resize_with(i + 1, Vec::new);
            }
            acc[i].push(num_str.parse().unwrap());
            acc
        });
    (cols, parse_operations(input))
}

fn parse_input_right_to_left(input: &str) -> (Vec<Vec<u64>>, Vec<Operation>) {
    let cols = input
        .lines()
        .take_while(|l| !l.starts_with(['*', '+']))
        .flat_map(|line| line.chars().enumerate())
        .filter_map(|(i, c)| c.to_digit(10).map(|d| (i, u64::from(d))))
        // Extract numbers column-wise
        // Columns -> [col1_num, col2_num, None, col3_num, ...]
        .fold(Vec::<Option<u64>>::new(), |mut acc, (i, d)| {
            if i >= acc.len() {
                acc.resize(i + 1, None);
            }
            match &mut acc[i] {
                Some(num) => *num = *num * 10 + d,
                None => acc[i] = Some(d),
            }
            acc
        })
        .into_iter()
        // Using None as a separator between groups of numbers
        .fold(vec![vec![]], |mut groups, opt_num| {
            match opt_num {
                Some(num) => groups.last_mut().unwrap().push(num),
                None => groups.push(Vec::new()),
            }
            groups
        });
    (cols, parse_operations(input))
}

fn parse_operations(input: &str) -> Vec<Operation> {
    input
        .lines()
        .last()
        .expect("Input should have at least one line")
        .split_whitespace()
        .map(Operation::from_str)
        .collect()
}

fn calculate(numbers: &[Vec<u64>], operations: &[Operation]) -> u64 {
    operations
        .iter()
        .zip(numbers.iter())
        .map(|(op, nums)| match op {
            Operation::Multiply => nums.iter().product::<u64>(),
            Operation::Add => nums.iter().sum(),
        })
        .sum()
}

pub fn calculate_columns(input: &str) -> u64 {
    let (numbers, operations) = parse_input_columns(input);
    calculate(&numbers, &operations)
}

pub fn calculate_right_to_left(input: &str) -> u64 {
    let (numbers, operations) = parse_input_right_to_left(input);
    calculate(&numbers, &operations)
}

pub fn main() {
    let input = std::fs::read_to_string("input/day06.txt").expect("Input file not found");
    let now = std::time::Instant::now();
    println!("PART 1 = {}", calculate_columns(&input));
    println!("PART 2 = {}", calculate_right_to_left(&input));
    println!("Execution time: {:?}", now.elapsed());
}
